### DISCUSSION TITLE
Select until the end/beginning of a function

### DISCUSSION THREAD

**Discussion by Tudyx** [helix-editor/helix#7024] [Ideas]
Currently if we are within a word and we use the `w` key, it will select from the cursor until the end of the word.
I thought it could be nice if we do the same with the tree sitter move "select next function", `]f`.

For instance, in the following function if the cursor is at the end of line1 and we press `]f`, then it will select line2 and line3.

```rust
fn foo() {
    println!("line1");
    println!("line2");
    println!("line3");
}
```

If you think it's a good idea and that it’s feasible, I could submit a PR implementing this.

**Comment by the-mikedavis** [comment]
This seems like it might be a nieche feature to me and usually solveable with `t` or `f`: in this case `t}` would select until the closing brace of the function. This would also take up a lot of keymap space. `]f`/`[f` already exists so a new binding would be necessary. Plus this would probably be implemented for all textobjects and there are currently 10, with PRs adding more.

**Comment by Tudyx** [comment]
Thanks for taking the time to reply!
> This seems like it might be a nieche feature to me and usually solveable with t or f: in this case t} would select until the closing brace of the function.
- One of the advantages of `]f` is that it will work wherever you are within the function. If you use `f}`/`t}` and you are inside a bracket inside the function, an if statement for instance, you will have to repeat `A-.` until you reach the end of the function.
- Another advantage of `]f` is that the user will only have to memorize one key combination and it will work for all languages. For instance, it will also work for Python which doesn't use brackets, where `f}`/`t}`  won't.

>  Plus this would probably be implemented for all textobjects and there are currently 10, with PRs adding more.

I agree if the community thinks this is useful we should go all the way down.

**Comment by pascalkuthe** [comment]
I this k you can already do this with `maf`. It puts the cursor at the end of the function. To put it at the start of the function you can flip the selection.

**Comment by Tudyx** [comment]
That's a great way to go to the end/beginning of a function, but what I meant is to **select** until the end/beginning of the function. Just selecting that part, not the whole function.

**Reply by sweetbbak** [reply]
``` vt} ```
visual + find_next_char + character... and then you can either use F f or t T to select to the end of a function from either direction.

**Reply by mweihs** [reply]
I think what would make things easier/possible would be "select until regex match". So we could type`v/^}`, meaning, select until the line beginning with "}" (= end of function).
Also "select until last position", `vC-o` would be nice?

**Reply by Tudyx** [reply]
> `vt}` visual + find_next_char + character... and then you can either use F f or t T to select to the end of a function from either direction.

Please see my [previous comment](https://github.com/helix-editor/helix/discussions/7024#discussioncomment-5881747) about why I think `[f`/`]f` would be better than `t}`
